table/compaction_merging_iterator.cc

```diff
@@ -0,0 +1,136 @@
+#include "table/compaction_merging_iterator.h"
+
+namespace ROCKSDB_NAMESPACE {
+void CompactionMergingIterator::SeekToFirst() {
+  minHeap_.clear();
+  status_ = Status::OK();
+  for (auto& child : children_) {
+    child.iter.SeekToFirst();
+    AddToMinHeapOrCheckStatus(&child);
+  }
+
+  for (size_t i = 0; i < range_tombstone_iters_.size(); ++i) {
+    if (range_tombstone_iters_[i]) {
+      range_tombstone_iters_[i]->SeekToFirst();
+      InsertRangeTombstoneAtLevel(i);
+    }
+  }
+
+  FindNextVisibleKey();
+  current_ = CurrentForward();
+}
+
+void CompactionMergingIterator::Seek(const Slice& target) {
+  minHeap_.clear();
+  status_ = Status::OK();
+  for (auto& child : children_) {
+    child.iter.Seek(target);
+    AddToMinHeapOrCheckStatus(&child);
+  }
+
+  ParsedInternalKey pik;
+  ParseInternalKey(target, &pik, false /* log_err_key */)
+      .PermitUncheckedError();
+  for (size_t i = 0; i < range_tombstone_iters_.size(); ++i) {
+    if (range_tombstone_iters_[i]) {
+      range_tombstone_iters_[i]->Seek(pik.user_key);
+      // For compaction, output keys should all be after seek target.
+      while (range_tombstone_iters_[i]->Valid() &&
+             comparator_->Compare(range_tombstone_iters_[i]->start_key(), pik) <
+                 0) {
+        range_tombstone_iters_[i]->Next();
+      }
+      InsertRangeTombstoneAtLevel(i);
+    }
+  }
+
+  FindNextVisibleKey();
+  current_ = CurrentForward();
+}
+
+void CompactionMergingIterator::Next() {
+  assert(Valid());
+  // For the heap modifications below to be correct, current_ must be the
+  // current top of the heap.
+  assert(current_ == CurrentForward());
+  // as the current points to the current record. move the iterator forward.
+  if (current_->type == HeapItem::ITERATOR) {
+    current_->iter.Next();
+    if (current_->iter.Valid()) {
+      // current is still valid after the Next() call above.  Call
+      // replace_top() to restore the heap property.  When the same child
+      // iterator yields a sequence of keys, this is cheap.
+      assert(current_->iter.status().ok());
+      minHeap_.replace_top(current_);
+    } else {
+      // current stopped being valid, remove it from the heap.
+      considerStatus(current_->iter.status());
+      minHeap_.pop();
+    }
+  } else {
+    assert(current_->type == HeapItem::DELETE_RANGE_START);
+    minHeap_.pop();
+    size_t level = current_->level;
+    assert(range_tombstone_iters_[level]);
+    range_tombstone_iters_[level]->Next();
+    InsertRangeTombstoneAtLevel(level);
+  }
+  FindNextVisibleKey();
+  current_ = CurrentForward();
+}
+
+void CompactionMergingIterator::FindNextVisibleKey() {
+  while (!minHeap_.empty() && minHeap_.top()->IsDeleteRangeSentinelKey()) {
+    HeapItem* current = minHeap_.top();
+    // range tombstone start keys from the same SSTable should have been
+    // exhausted
+    assert(!range_tombstone_iters_[current->level] ||
+           !range_tombstone_iters_[current->level]->Valid());
```

Comment:
I got lost here until realizing `minHeap_.top()->IsDeleteRangeSentinelKey()` returns false for range tombstone start keys. It's kind of confusing because CompactionMergingIterator::IsDeleteRangeSentinelKey() would return true for such keys. I'm still not 100% sure but either way some more comments on this could help.